Pymc3 Plot & Diagnostics & Arviz Dependency

CloudChaoszero opened the pull request:
# Description

This PR  starts the (Breakchange) removal of PyMC3 Plots and Diagnostics process referred in #4371, and relying on referred functionality from Arviz. See an example from the quote below

> Plots and diagnostics are still accessible via PyMC3 but are actually implemented by ArviZ now. Although this was a good call when ArviZ was still nascent, now this causes confusion for new comers and overhead for devs, who have to mirror the functions in ArviZ, which is more work when ArviZ adds or changes functions (e.g pm.hpd vs. az.hdi).

# The Implementation Process

## Add Deprecation warning to one remaining (plot) function
Add a `DeprecateWarning` to function `plot_posterior_predictive_glm`

## ~V4~ 3.11  implementation

1. Remove Diagnostics (pymc3/stats), incrementally
2. Remove Plots (pymc3/plots), incrementally 
3. Test
4. Ensure Jupyter Notebook examples in [pymc3-examples](https://github.com/pymc-devs/pymc-examples) is updated as well

# Notes 

+ [X] what are the (breaking) changes that this PR makes?

+ [X] important background, or details about the implementation
+ [X] are the changes—especially new features—covered by tests and docstrings?
+ [ ] consider adding/updating relevant example notebooks
+ [ ] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md

CloudChaoszero committed:
:tada: Start removing Diagnostics & Plots in PyMC3 development

:fire: Remove arviz plots
* Remove directly imported arviz plots used in pymc3 plots

:fire: Remove all plots from PyMC3 Plots module

:fire: Remove PyMC3 plots references in Docs

:art: Mention Plotting & Diagnostics in API page and remove plots reference in __init__.py

:rewind: Revert posterior_plot function, test, and docs

:art: Add deprecation warning to posterior_plot function

:art: Add context on plot import and import back into __init__.py

:pencil2: Add warning and details of posterior_plot added

Update docs/source/api/plots.rst

Co-authored-by: Alexandre ANDORRA <andorra.alexandre@gmail.com>

Update docs/source/api/plots.rst

Co-authored-by: Alexandre ANDORRA <andorra.alexandre@gmail.com>

Update pymc3/plots/__init__.py

Co-authored-by: Alexandre ANDORRA <andorra.alexandre@gmail.com>

Update pymc3/plots/__init__.py

Co-authored-by: Alexandre ANDORRA <andorra.alexandre@gmail.com>

:pencil2: Update docs to add stats.rst details

:pencil2: Minor docs notation for posterioplot function(s)

:pencil: Add breakline before docstring title

CloudChaoszero commented:
Note to self: Is `v4` the base branch?

twiecki commented:
Thanks @CloudChaoszero! I think we should make sure to also change the `pymc-examples` so that people can still run those fine. That should probably happen before we release with this change.

CloudChaoszero commented:
> Thanks @CloudChaoszero! I think we should make sure to also change the `pymc-examples` so that people can still run those fine. That should probably happen before we release with this change.

Sounds great. I'll proceed with that in the  `pymc-examples`, and after update this branch.

OriolAbril commented on docs/source/api.rst:
```diff
@@ -17,8 +17,6 @@ API Reference
    api/glm
    api/gp
-   api/plots
-   api/stats
```
I would probably add a line after the API toctree saying that PyMC3 delegates plots and stats to ArviZ and a link to https://arviz-devs.github.io/arviz/. Not sure why but even though pymc docs have only had links to ArviZ for quite some time now, I still see plenty of people using `pm.summary`, `pm.traceplot`... which may be confused by plots and stats sections disappearing from the docs.

CloudChaoszero replied:
Sounds good @OriolAbril . How does this look?
I added a separate section, mentioning plotting and diagnostics.

<img width="313" alt="Screen Shot 2021-01-04 at 12 37 27 AM" src="https://user-images.githubusercontent.com/19514362/103517341-cd003a00-4e26-11eb-9fbd-efddbd6009bd.png">


Also, to the part of people using `pm.summary` or `pm.traceplot`, from what I understand, the initial dev for this repo didn't have arviz's advantages of being mature than what it is now. 

However, that is just paraphrasing what I read 🤷‍♂️  lol

CloudChaoszero replied:
Also @OriolAbril , do you know which branch is the correct branch for the `v4` release?

OriolAbril replied:
This looks great, thanks!

AlexAndorra replied:
I think it's named `v4`, right @twiecki ? So this PR should probably be based on this branch

michaelosthege replied:
A `v4` branch (not the one we currently have in the repo) will be used for `v4` development, but we're not there yet.

I think we should merge this PR with `master` instead of trying to keep it in sync for so long.

OriolAbril reviewed (commented):
Everything looks good so far, I only have a minor suggestion about the api page. I'm looking forward to have the example notebooks updated.

OriolAbril commented on pymc3/tests/test_examples.py:
```diff
@@ -212,7 +213,7 @@ def test_disaster_model_missing(self):
             # Use slice sampler for means (other variables auto-selected)
             step = pm.Slice([model.early_mean_log__, model.late_mean_log__])
             tr = pm.sample(500, tune=50, start=start, step=step, chains=2)
-            pm.summary(tr)
+            az.summary(tr)
```
Just had a revelation that now that we are updating this, we could also start using `return_inferencedata=True` when possible, here for example we could do:

```
    tr = pm.sample(..., return_inferencedata=True)
az.summary(tr)
```

once we are using an inferencedata it is not necessary to call `az.summary` from inside a model context. Does this sound useful to prepare people for `return_inferencedata=True` becoming the default?

AlexAndorra replied:
It would definitely be useful! Although once this does become the default, maybe we'll have to remove the kwarg, which would be cumbersome to do this change two times 🤔

michaelosthege replied:
We should change all tests to explicitly pass `return_inferencedata` in one PR.

Only then we can swap the default, which I think we should do with v4. After that we can, but don't need to take out the `return_inferencedata=True` appearances in one PR.

AlexAndorra replied:
Sounds good @michaelosthege, let's do that

MarcoGorelli commented on pymc3/plots/posteriorplot.py:
```diff
@@ -1,71 +0,0 @@
-    from arviz.data.inference_data import InferenceData
-
-
-def plot_posterior_predictive_glm(
```
this exact function isn't (yet) available in arviz and so perhaps shouldn't be removed just yet (this is already mentioned in the issue, just repeating it here for verbosity in case it's overlooked)

michaelosthege replied:
Yes, looks like the discussion in https://github.com/pymc-devs/pymc3/issues/4371 settled on keeping just `plot_posterior_predictive_glm` around, but let's add a `DeprecationWarning`.

CloudChaoszero replied:
Sounds great, I'll add that with the following:
```python
    warnings.warn(
        "The `plot_posterior_predictive_glm` function will migrate to Arviz in a future release. "
        "\nKeep up to date with `ArviZ <https://arviz-devs.github.io/arviz/>`_ for future updates.",
        DeprecationWarning,
    )
```

CloudChaoszero replied:
Sounds good!

michaelosthege reviewed (changes requested):
Hi @CloudChaoszero ,
I think this PR should move forward and get merged before the `3.11.0` release. We should keep just the `plot_posterior_predictive_glm` (and its tests) around because there is no replacement for it.

I think it's fair to remove the other plotting and stats functions, because we advertised ArviZ for a long time and even though they did not have deprecation warnings, their replacement is just a minor inconvenience for the users. `3.11` will bring much more severe breaking changes than this.

CloudChaoszero commented:
Thanks for the feedback y'all! 😄 

Alright, so I implemented the following, based on reviews:

## 3.11
This PR is against `master` for 3.11 release.

## Documentation
@OriolAbril 
Added the `plots` section back into the documentation. Also, I changed the description to focus on arviz, but PyMC3 does offer `plot_posterior_predictive_glm `, seen below:

<img width="709" alt="Screen Shot 2021-01-17 at 2 41 30 AM" src="https://user-images.githubusercontent.com/19514362/104838096-9b975d80-586d-11eb-8116-8e523dacc924.png">

## Keeping the plot_posterior_predictive_glm Function

@michaelosthege  I went ahead and added a warning to `plot_posterior_predictive_glm`

```python
    warnings.warn(
        "The `plot_posterior_predictive_glm` function will migrate to Arviz in a future release. "
        "\nKeep up to date with `ArviZ <https://arviz-devs.github.io/arviz/>`_ for future updates.",
        DeprecationWarning,
    )
```
_______
@AlexAndorra Question: Are there any other Arviz diagnostics dependencies I should consider, outside of the plots topic?

AlexAndorra commented on docs/source/about.rst:
```diff
@@ -88,7 +88,7 @@ of PyMC. PyMC 2.2 was released in April 2012. It contained numerous bugfixes and
 optimizations, as well as a few new features, including improved output
 plotting, csv table output, improved imputation syntax, and posterior
 predictive check plots. PyMC 2.3 was released on October 31, 2013. It included
-Python 3 compatibility, improved summary plots, and some important bug fixes.
```
This shouldn't be modified, as it's just the history of PyMC

CloudChaoszero replied:
Ah, will revert!

AlexAndorra commented on docs/source/api/plots.rst:
```diff
@@ -9,15 +9,10 @@ Plots are delegated to the
 For plots, ``pymc3.<function>`` are now aliases
-for ArviZ functions. Thus, the links below will redirect you to
-ArviZ docs:
+for ArviZ functions.
```
```suggestion
Refer to its documentation to use the plotting functions directly.
```

AlexAndorra commented on docs/source/api/plots.rst:
```diff
@@ -9,15 +9,10 @@ Plots are delegated to the
 library, a general purpose library for
 "exploratory analysis of Bayesian models."
 For plots, ``pymc3.<function>`` are now aliases
```
```suggestion
```
You should also remove `.. currentmodule:: pymc3.plots` from this file

AlexAndorra commented on docs/source/api/stats.rst:
```diff
@@ -1,25 +0,0 @@
-"exploratory analysis of Bayesian models."
-For statistics and diagnostics, ``pymc3.<function>`` are now aliases
-for ArviZ functions. Thus, the links below will redirect you to
-ArviZ docs:
```
I think we should keep these lines around, at least for 3.11, to ease the transition, just like we're doing for `plots.rst`. So, something like that:
```raw
*****
Stats
*****
Statistics and diagnostics are delegated to the
`ArviZ <https://arviz-devs.github.io/arviz/index.html>`_.
library, a general purpose library for
"exploratory analysis of Bayesian models."
Refer to its documentation to use the diagnostics functions directly.
```

AlexAndorra commented on pymc3/plots/__init__.py:
```diff
@@ -14,108 +14,19 @@
-"exploratory analysis of Bayesian models." See https://arviz-devs.github.io/arviz/
+Plots are delegated to the `ArviZ <https://arviz-devs.github.io/arviz/>`_ library, a general purpose library for
+"exploratory analysis of Bayesian models." For more details, see https://arviz-devs.github.io/arviz/
 for details on plots.
```
```suggestion
```

AlexAndorra commented on pymc3/plots/__init__.py:
```diff
@@ -14,108 +14,19 @@
 for details on plots.
+
+In one case, PyMC3 does preserve the `plot_posterior_predictive_glm` to implement a combined
+implementation of both PyMC3 and Arviz to predictively pushforward samples.
```
```suggestion
Only the `plot_posterior_predictive_glm` is kept in the PyMC code base for now,  but it'll will move to ArviZ once the latter adds features for regression plot.
```

AlexAndorra reviewed (changes requested):
Thanks @CloudChaoszero, starting to look good!
I added a few comments and suggestions below to make it ready to merge 😉

michaelosthege commented:
@CloudChaoszero please try to bundle the commits. Every commit that is pushed causes the entire test pipeline to run, resulting in a lot of unnecessary compute.

When committing suggested changes from Github use the "add suggestion to batch" and when committing locally, try to push all commits at the same time - unless you want to trigger the CI for some specific reason.

Many currently running CI jobs can probably be cancelled..

CloudChaoszero commented:
>  please try to bundle the commits. Every commit that is pushed causes the entire test pipeline to run, resulting in a lot of unnecessary compute.....

@michaelosthege  Whoops, pardon! 😅  Thanks for the additional details.
_____
@AlexAndorra Updated the PR. I am (slowly) also working on [pymc-example PR](https://github.com/pymc-devs/pymc-examples/pull/16) to update docs, regarding this change.

MarcoGorelli commented on pymc3/plots/__init__.py:
```diff
@@ -1,4 +1,4 @@
-#   Copyright 2020 The PyMC Developers
+#   Copyright 2021 The PyMC Developers
```
I don't think this needs to change (and if it does, I think the change needs doing for the whole codebase?)

AlexAndorra replied:
Good point, let's not change that here then

MarcoGorelli commented on pymc3/plots/__init__.py:
```diff
@@ -14,108 +14,17 @@
-"exploratory analysis of Bayesian models." See https://arviz-devs.github.io/arviz/
-for details on plots.
+Plots are delegated to the `ArviZ <https://arviz-devs.github.io/arviz/>`_ library, a general purpose library for
+"exploratory analysis of Bayesian models." For more details, see https://arviz-devs.github.io/arviz/
```
```suggestion
"exploratory analysis of Bayesian models". For more details, see https://arviz-devs.github.io/arviz/ .
```

OriolAbril replied:
I would actually remove the `"` in exploratory analysis of Bayesian models.

MarcoGorelli commented on pymc3/plots/__init__.py:
```diff
@@ -14,108 +14,17 @@
+Plots are delegated to the `ArviZ <https://arviz-devs.github.io/arviz/>`_ library, a general purpose library for
+"exploratory analysis of Bayesian models." For more details, see https://arviz-devs.github.io/arviz/
+
+Only the `plot_posterior_predictive_glm` is kept in the PyMC code base for now,  but it'll will move to ArviZ once the latter adds features for regression plot.
```
```suggestion
Only `plot_posterior_predictive_glm` is kept in the PyMC code base for now, but it will move to ArviZ once the latter adds features for regression plots.
```

MarcoGorelli reviewed (approved):
Nice - barring a couple of nitpicks, this looks good to me

AlexAndorra commented on docs/source/api/plots.rst:
```diff
@@ -8,16 +8,10 @@ Plots are delegated to the
-- :func:`pymc3.densityplot <arviz:arviz.plot_density>`
-- :func:`pymc3.pairplot <arviz:arviz.plot_pair>`
+.. warning::
+    The `plot_posterior_predictive_glm` function will removed in a future PyMC3 release.
```
```suggestion
    The `plot_posterior_predictive_glm` function will be removed in a future PyMC3 release.
```

AlexAndorra reviewed (changes requested):
Thanks for the quick update @CloudChaoszero ! Still a few typos to iron out

OriolAbril commented on docs/source/api/plots.rst:
```diff
@@ -8,16 +8,10 @@ Plots are delegated to the
-- :func:`pymc3.kdeplot <arviz:arviz.plot_kde>`
-- :func:`pymc3.densityplot <arviz:arviz.plot_density>`
-- :func:`pymc3.pairplot <arviz:arviz.plot_pair>`
+.. warning::
```
I would not use an sphinx warning here but inside its api docs. 

```
"""Plot posterior predictive of a linear model.

    Parameters
    ----------
    trace : arviz.InferenceData or MultiTrace
        Output of pm.sample()
    eval : array_like
        Array over which to evaluate lm
    lm : function, default: linear function
        Function mapping parameters at different points
        to their respective outputs.
            input: point, sample
            output: estimated value
    samples : int, default=30
        How many posterior samples to draw.
    kwargs : mapping, optional
        Additional keyword arguments are passed to ``matplotlib.pyplot.plot()``.

    Warnings
    --------
    The `plot_posterior_predictive_glm` function will removed in a future PyMC3 release.
    """
```

something like that.

CloudChaoszero replied:
Sounds great @OriolAbril I added it. However, the Warning did not show in my local version. 🤷 

<img width="641" alt="Screen Shot 2021-01-17 at 10 21 45 PM" src="https://user-images.githubusercontent.com/19514362/104879344-86c0d580-5912-11eb-9ee1-f7ce1bf4579e.png">

CloudChaoszero commented:
Sounds great @AlexAndorra. I went ahead and also updated the plots docs to have the warning be withing the `plot_posterior_predictive_glm `, from recommendation from @ OriolAbril.

Example
<img width="641" alt="Screen Shot 2021-01-17 at 10 21 45 PM" src="https://user-images.githubusercontent.com/19514362/104879724-4150d800-5913-11eb-9730-70dbbf8e9f54.png">

OriolAbril commented on pymc3/plots/posteriorplot.py:
```diff
@@ -33,20 +35,32 @@ def plot_posterior_predictive_glm(
-        samples: int <default=30>
-            How many posterior samples to draw.
-    Additional keyword arguments are passed to pylab.plot().
+    Parameters
```
I think there should be an empty line before the section titles.

OriolAbril reviewed (approved):
I take the warning in the api docs is working with the current docstring? The two images without the warning are the same and don't match the docstring text. Other than that everything looks great

AlexAndorra commented on pymc3/plots/posteriorplot.py:
```diff
@@ -33,20 +35,33 @@ def plot_posterior_predictive_glm(
+
+    Warnings
+    --------
+    The `plot_posterior_predictive_glm` function will removed in a future PyMC3 release.
```
```suggestion
    The `plot_posterior_predictive_glm` function will be removed in a future PyMC3 release.
```

AlexAndorra commented on docs/source/api/plots.rst:
```diff
@@ -8,16 +8,9 @@ Plots are delegated to the
-- :func:`pymc3.pairplot <arviz:arviz.plot_pair>`
+.. currentmodule:: pymc3.plots.posteriorplot
+
+.. automodule:: pymc3.plots.posteriorplot
```
Are both those lines needed here, or only one of them?

CloudChaoszero replied:
Hmm, I think 1 of those. I removed
`.. currentmodule:: pymc3.plots.posteriorplot` for the time being

AlexAndorra reviewed (changes requested):
Thanks @CloudChaoszero ! Found one last typo. And does the warning display in the plot docs now?

CloudChaoszero committed:
:pencil2: small typo and remove summary in plots.rst

CloudChaoszero commented:
>  And does the warning display in the plot docs now?

Done! And @AlexAndorra I believe not. 
However, that was in my local build. Unsure if the same will be displayed.
* Quite odd that the `Warnings` section did not show.  We do have the DeprecateWarning in the function though... 🤷 haha

<img width="735" alt="Screen Shot 2021-01-18 at 2 24 38 AM" src="https://user-images.githubusercontent.com/19514362/104903032-5be77900-5934-11eb-8984-80616fb7bf8f.png">

AlexAndorra commented:
Yeah but we definitely need this warning in the docs as well. And I don't see why it would not work for you locally but work when we recompile the new docs. What do you think @OriolAbril ?
The last thing we can do is add back the warning in the docs text (as well as keep it in the function docstrings).

MarcoGorelli commented:
@CloudChaoszero did you do `make clean` before rebuilding? It appears fine for me

![image](https://user-images.githubusercontent.com/33491632/104933862-54aa8480-59a1-11eb-98cf-f7614777af80.png)

CloudChaoszero commented:
> @CloudChaoszero did you do `make clean` before rebuilding? It appears fine for me

@MarcoGorelli Hmm, maybe not. But yeah, that looks good to me! 👏  thanks for verifying.

CC: @AlexAndorra

AlexAndorra reviewed (approved):
Aaaah, perfect then, thanks for chiming in with some doc magic @MarcoGorelli 🧙‍♂️ 
This is good to merge now 🥳 Thanks @CloudChaoszero !
